Update search view updating

m-rosenbaum opened the pull request:
This updates the DOM replacement function to a bit more decomposed format. After the search button is pressed, the loading state is replaced with named divs for us to replace information from the `fetch_all_data` response in.

@mmbarros-git, I'm adding this branch in since it contains the main refactor. Once this is confirmed, I will add in functions to take the API response and fill each of these divs as relevant with the information from the `fetch_all_data` response.

Of note in this lines 89-115 are a lot of DOM modifications that are reasonable static. Do you think this should come from a template like I did with filters HTML?

m-rosenbaum commented:
@mmbarros-git Based on what we've talked about on Wednesday, this PR now includes the following updates. It should be ready for your review. I'm also adding in James as well (comment to follow).

- A loading bar is added to the search div. (Should this be over the whole page). I've included some basic styling here so it matches the Bulma color scheme.
- This now takes the response object and starts to populate the search div with the information from the response object.

It does not make the `fetch_*` requests that we talked about today or include any of the visual clean-up in our [issue for visual bugs](https://github.com/uchicago-capp-30320/apt-plus/issues/50).

m-rosenbaum commented:
@jamesturk Thank you for agreeing to look at our JavaScript. 

**Request**
We talked briefly about getting feedback on syntactic approaches and function decomposition/abstraction as Maida and I are less familiar with JavaScript best practices. 

Any feedback on better approaches would be extremely helpful here as we are going to b writing a bit more frontend modification code.

**Context**
This commit covers a refactor of the functionality you saw during our demo on May 8th. As a reminder, this branch can take an address in Hyde Park, geocode it, and place it on the map. The first API request takes ~1.8 seconds to complete.

Because of that, our frontend code (1) adds a loading screen to inform the user that something is happening and (2) then modifies the state of the view to switch to our "search mode" which updates our left search panel with controls for the map. To do (2), it first adds a CSS-styled divs that we will fill with responses from our other dataset-specific APIs (`fetch_inspections`, `fetch_amenities`, etc.). 

All of the code that does this modification is located in the:

- `static/js/address_submission.js` - contains all of the functions to modify the DOM
- `static/js/home.css` - contains all the custom CSS classes used
- `templates/address_search` - contains a template for the filter button panel that we load into the JS

There are a number of visual design issues in the HTML. These are tracked in a [single issue](https://github.com/uchicago-capp-30320/apt-plus/issues/50) and are not in scope for this branch.

mmbarros-git commented:
Thank you so much, @m-rosenbaum, for all your hard work on this. Also, I really liked how the loading spinner turned out!

A couple of notes:

- When I try searching for an address from the second page (i.e., while already on an apartment detail view), it doesn’t seem to be working. I recall submitting that functionality previously and that it had been merged, but it looks like the logic has changed with the refactor, and this feature is not working when I test it locally.

- One small issue to address in the error handling is that the loading spinner continues spinning even when an error is thrown (e.g., if the address can't be geocoded). Ideally, the spinner should stop when an error occurs. I can do that with the error popups functionality. 

- Just a minor UI suggestion: it might be helpful to display an additional line under the street address showing “Chicago, IL, [zip code]” to provide more context. I know my original template didn’t include that, but it could improve clarity for users.

mmbarros-git commented:
Hi @m-rosenbaum I just added the new error popups in this branch. 

In the end, I was convinced by this style from Bulma: https://bulma.io/documentation/form/general/#complete-form-example. Like you suggested in the other branch. I think it looks cleaner and clearer for users, but if you’d prefer the previous version I had, it’s easy to swap back; it’s just styling.

Note: I did not include the proposed fix for the search button on the second page; however, I did use it temporarily to assist with debugging and ensure the error handling flow worked properly.

Also updated address_submission.js to follow what we discussed the other day:
- Errors are shown before switching to the second page.
- Spinner stops even when there’s an error.
- The error message clears when the user starts typing again.

mmbarros-git commented:
One note: at the moment, the code assumes that violations/inspections data will always be returned in the fetch_all_data response. If that section is missing or empty, there’s no fallback behavior yet, this could lead to UI issues or blank placeholders. We may want to add a conditional check or message for those cases in the next step.

m-rosenbaum commented:
Thanks for these! I appreciate you looking through these and adding in the error messages. Would it make sense to combine the error messages and my utility functions into a util.js?

> * When I try searching for an address from the second page (i.e., while already on an apartment detail view), it doesn’t seem to be working. I recall submitting that functionality previously and that it had been merged, but it looks like the logic has changed with the refactor, and this feature is not working when I test it locally.
> * One small issue to address in the error handling is that the loading spinner continues spinning even when an error is thrown (e.g., if the address can't be geocoded). Ideally, the spinner should stop when an error occurs. I can do that with the error popups functionality.

Thanks for handling these! I also updated the initial code to change content to new skeletons whenever the address button is pressed a second time.

> One note: at the moment, the code assumes that violations/inspections data will always be returned in the fetch_all_data response. If that section is missing or empty, there’s no fallback behavior yet, this could lead to UI issues or blank placeholders. We may want to add a conditional check or message for those cases in the next step.

Good thought. @whiskey0504 what would happen if a building isn't inspected? Under what conditions would a valid address not return inspections?

My loose understanding of Chicago building laws is that inspections occur annually for rental buildings and a new rental building would be required to be inspected prior to renting. Would the only cases where we wouldn't receive data be (i) API errors we could monitor, (ii) data errors on CHA's side in publications, and (iii) errors in our loading of data?

(i) seems like it could just be an error pop-up saying we're having issues retrieving data, but (ii) & (iii) seem harder to treat.

> * Just a minor UI suggestion: it might be helpful to display an additional line under the street address showing “Chicago, IL, [zip code]” to provide more context. I know my original template didn’t include that, but it could improve clarity for users.

Added!

mmbarros-git commented:
> Would it make sense to combine the error messages and my utility functions into a util.js?

Yes, that makes sense, error handling is definitely a reusable function, so it fits well in a util.js file. Who would you prefer to make that commit? I didn’t see a util.js file in the repo yet.

m-rosenbaum commented:
> > Would it make sense to combine the error messages and my utility functions into a util.js?
> 
> Yes, that makes sense, error handling is definitely a reusable function, so it fits well in a util.js file. Who would you prefer to make that commit? I didn’t see a util.js file in the repo yet.

If you are updating the error/warning messaging function feel free to make it then!

jamesturk reviewed (commented):
This looks very good overall- there are (as you've likely discovered) a lot of different styles of writing this kind of code in JS but this is a very readable choice which is what I'd prioritize on a team newer to JS.

One small suggestion:

```
    toggleLoadingWheel(); //Ensure spinner is removed even on failure
``` 

This is a good use of a `finally` block, like in Python, these blocks execute after the try/except no matter which branch was taken.

jamesturk commented:
The most contentious portion here is the reliance on createElement, this is where a lot of people either use a JS templating library or a component-based framework like React. 

There is no "right" way though, and the approach you chose is great for a team that isn't trying to invest heavily in JS (as in having 2-3 team members that are primarily JS authors that can handle the complexity/tooling needed for those frameworks).

mmbarros-git commented:
Hi @m-rosenbaum  
I just checked this and the button is now working, but the problem is that the bullets from the first page are not disappearing, son on the second page appear all the new things and also the bullets from the first page. 

> Thanks for handling these! I also updated the initial code to change content to new skeletons whenever the address button is pressed a second time.

mmbarros-git commented:
> Would it make sense to combine the error messages and my utility functions into a util.js?

mmbarros-git commented:
@jamesturk Thank you so much for reviewing the code!
I really appreciate the feedback. I've implemented your suggestion and moved the spinner logic into a finally block as recommended.
> 
> One small suggestion:
> 
> ```
>     toggleLoadingWheel(); //Ensure spinner is removed even on failure
> ```
> 
> This is a good use of a `finally` block, like in Python, these blocks execute after the try/except no matter which branch was taken.

m-rosenbaum commented:
Thanks for catching that!

I deleted that when making the final fix and didn't test it. I'm very excited for us to have the playwright tests integrated at sme point.

It's now working normally with the fix I just committed.

> Hi @m-rosenbaum I just checked this and the button is now working, but the problem is that the bullets from the first page are not disappearing, son on the second page appear all the new things and also the bullets from the first page.
> 
> > Thanks for handling these! I also updated the initial code to change content to new skeletons whenever the address button is pressed a second time.

mmbarros-git commented:
@m-rosenbaum  I already checked your update and it works ok for me!

I just created the utils.js with all the necessary for your final check and merge. Thanks!!
